middleware.go

```diff
@@ -0,0 +1,92 @@
+package jwt
+
+import (
+	"github.com/Digital-Insight-Technologies-Ltd/gin-jwt-cognito/models"
+	"net/http"
+
+	"github.com/gin-gonic/gin"
+	"github.com/go-errors/errors"
+	"github.com/golang-jwt/jwt/v5"
+	"github.com/google/uuid"
+)
+
+// UserClaimsKey is the key for user claims in our context
+const UserClaimsKey = "userClaims"
+
+// CognitoClaimsMiddleware is a middleware for Gin that extracts user claims from
+// AWS Cognito.
+func CognitoClaimsMiddleware() gin.HandlerFunc {
+	return func(c *gin.Context) {
+		claims, err := parseClaims(c.GetHeader("Authorization")[7:])
+		if err != nil {
+			c.AbortWithStatusJSON(http.StatusUnauthorized, gin.H{"error": "Invalid user claims (OIDC: stage 1)"})
+			return
+		}
+
+		userCtx, err := constructUserContext(claims)
+
+		if err != nil {
+			c.AbortWithStatusJSON(http.StatusUnauthorized, gin.H{"error": "Invalid user claims (OIDC: stage 2)"})
+			return
+		}
+
+		// Add the claims to Gin's context
+		c.Set(UserClaimsKey, userCtx)
+
+		// Proceed with the middleware chain
+		c.Next()
```

Comment:
Consider adding context cancellation or timeout handling to ensure the middleware does not hang indefinitely in case of external service delays. [medium]